src/vm_pages.rs

```diff
@@ -151,31 +158,201 @@ impl TlbTracker {
     }
 }
 
+/// Types of regions in a VM's guest physical address space.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum VmRegionType {
+    /// Regions of private memory that have been fully populated during VM creation.
+    PreMapped,
+    /// Placeholder region type to reserve a range of guest physical address space while the region
+    /// is mapped.
+    Reserved,
+}
+
+/// A contiguous region of guest physical address space.
+struct VmRegion {
+    start: GuestPageAddr,
+    end: GuestPageAddr,
+    region_type: VmRegionType,
+}
+
+/// Represents a reserved region in `VmRegionList` that has yet to be committed. Removes the reserved
+/// region if dropped without being passed to `VmRegionList::commit()`.
+pub struct VmReservedRegion<'a> {
+    owner: &'a VmRegionList,
+    start: GuestPageAddr,
+    region_type: VmRegionType,
+}
+
+impl<'a> VmReservedRegion<'a> {
+    /// Creates a new reserved region holder in `owner` for the region beginning at `start` with
+    /// type `region_type`.
+    fn new(owner: &'a VmRegionList, start: GuestPageAddr, region_type: VmRegionType) -> Self {
+        Self {
+            owner,
+            start,
+            region_type,
+        }
+    }
+}
+
+impl<'a> Drop for VmReservedRegion<'a> {
+    fn drop(&mut self) {
+        let mut regions = self.owner.regions.lock();
+        regions.retain(|r| r.start != self.start);
+    }
+}
+
+/// The regions of guest physical address space for a VM. Used to track which parts of the address
+/// space are designated for a particular purpose. The region list is created during VM initialization
+/// and remains static after the VM is finalized.
+pub struct VmRegionList {
+    regions: Mutex<PageVec<VmRegion>>,
+}
+
+impl VmRegionList {
+    /// Creates a new `VmRegionList` using `pages` as the backing store.
+    pub fn new(pages: SequentialPages<InternalClean>, page_tracker: PageTracker) -> Self {
+        Self {
+            regions: Mutex::new(PageVec::new(pages, page_tracker)),
+        }
+    }
+
+    /// Reserves a region at [`start`, `end`) of type `region_type`, returning a `VmReservedRegion`
+    /// object if it does not conflict with any existing regions. After performing any actions
+    /// necessary to populate the region (e.g. filling in page tables), the `VmReservedRegion` should
+    /// be passed to `commit()` to commit it to the `VmRegionList`. If the `VmReservedRegion` is
+    /// dropped before it is committed, the reservation is removed.
+    pub fn prepare(
```

Comment:
pub needed? Seems to only be used in this file.

Reply:
Removed.